### PR TITLE
Leveling, Player Info and Game Update XP Placeholders

### PR DESCRIPTION
**Pull request by Zffu**
Hi, i have added some placeholders and changed some there is a list of things i have done

**Info Component Changes**
- Added the %travel% placeholder (Traveling time to get to the beacon parsed in duration)
- Added the %travel_raw% placeholder (Traveling time to get to the beacon in seconds)

**Game Update Component Changes**
- Added the %pc_formatted% placeholder (the %pc% placeholder but parsed)

**Leveling Component Changes**
- Added the %grindtime% placeholder (Griding time to get to the next level)
- Added the %grindtime_raw% placeholder (Griding time to get to the next level in seconds)
- Added the %max_formated% placeholder
- Added the %actual_formated% placeholder

**Util Functions**
- Added a duration parser (DurationintegerToShortString() function)
- Added a function to get the traveling time (getTravelingTime() function)

Also i would like to be a Wynntils Developper if its possible

### CODE REVIEW

File: src/main/java/com/wynntils/modules/utilities/instances/InfoFormatter.java [hunk 1]
```diff
@@ -134,6 +144,8 @@ public InfoFormatter() {
                 Integer.toString(PlayerInfo.get(CharacterData.class).getMaxHealth()),
                 "health_max");
 
+
+
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/instances/InfoFormatter.java]: Why random new spaces?

File: src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java [hunk 2]
```diff
@@ -41,6 +42,31 @@ public static void update() {
...
+        }
+
+
+    }
+
+
+
+
+
+
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java]: Random spaces

File: src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java [hunk 3]
```diff
@@ -41,6 +42,31 @@ public static void update() {
...
+     * Get Excepted Grind Time for leveling up
+     * @return The Grind Time for Leveling Up in Minutes
+     */
+    public static int getLevelingGrindTime(){
+        CharacterData p = PlayerInfo.get(CharacterData.class);
+        int perminute = getXpPerMinute();
+        int required = p.getXpNeededToLevelUp() - p.getCurrentXP();
+
+
+        if((!(required == 0)) && (!(perminute == 0))){
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java]: You can simplify this :)

**Comment by Zffu** [comment]
If you want me to add something please tell me

File: src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java (hunk 3 above)
Comment (kristofbolyai) [reply src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java]: You also only need to check if perminute is 0. You can divide zero by x and not get an exception.

File: src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java [hunk 4]
```diff
@@ -20,4 +22,23 @@ public static double getCurrentSpeed() {
         return (MathHelper.sqrt((distX * distX) + (distZ * distZ))) * 20d;
     }
 
+    public static int getTravelingTime(){
+        Location compass = CompassManager.getCompassLocation();
+        if(!(compass == null)){
+            Location playerPos = new Location(McIf.player());
+            double distance = compass.distance(playerPos);
+            double bps = getCurrentSpeed();
+            if(!(bps == 0)){
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java]: I am not sure why you don't use `!=`

File: src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java [hunk 5]
```diff
@@ -20,4 +22,23 @@ public static double getCurrentSpeed() {
         return (MathHelper.sqrt((distX * distX) + (distZ * distZ))) * 20d;
     }
 
+    public static int getTravelingTime(){
+        Location compass = CompassManager.getCompassLocation();
+        if(!(compass == null)){
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java]: !=

File: src/main/java/com/wynntils/modules/utilities/instances/InfoFormatter.java (hunk 1 above)
Comment (DevScyu) [inline src/main/java/com/wynntils/modules/utilities/instances/InfoFormatter.java]: Please remove empty lines

File: src/main/java/com/wynntils/core/utils/StringUtils.java [hunk 6]
```diff
@@ -453,6 +453,20 @@ public static String integerToShortString(int count) {
         return String.format("%s%c", value, "kMBTPE".charAt(exp - 1));
     }
 
+    public static String DurationintegerToShortString(int count){
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/core/utils/StringUtils.java]: I am pretty sure there is something like this somewhere in the code-base. Anyways I am sure I can't tell you where. However the method naming is wrong, use lowerCamelCase.

File: src/main/java/com/wynntils/core/utils/StringUtils.java [hunk 7]
```diff
@@ -453,6 +453,20 @@ public static String integerToShortString(int count) {
...
+        if(count < 60){
+            return "" + count + "s";
+        }
+        else {
+            if(count > 3600){
+                return "" + count / 3600 + "h";
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/core/utils/StringUtils.java]: String.valueOf() so you don't need to "" + ?

File: src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java (hunk 2 above)
Comment (DevScyu) [inline src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java]: Please remove empty lines

File: src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java [hunk 8]
```diff
@@ -134,7 +134,10 @@ public void onClientTick(TickEvent.ClientTickEvent e) {
                                         .replace("%xc%", Integer.toString(data.getCurrentXP() - oldxp))
                                         .replace("%po%", oldxppercent)
                                         .replace("%pn%", data.getCurrentXPAsPercentage())
-                                        .replace("%pc%", df.format(xpchange)));
+                                        .replace("%pc%", com.wynntils.core.utils.StringUtils.integerToShortString((int) xpchange))
+                                        .replace("%pc_raw%", df.format(xpchange)));
+
+
```
Comment (DevScyu) [inline src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java]: Please remove empty lines

File: src/main/java/com/wynntils/core/framework/instances/data/CharacterData.java [hunk 9]
```diff
@@ -60,6 +60,8 @@ public void updatePlayerClass(ClassType newClass, boolean newClassIsReskinned) {
         reskinned = newClassIsReskinned;
     }
 
+
+
```
Comment (DevScyu) [inline src/main/java/com/wynntils/core/framework/instances/data/CharacterData.java]: Please remove empty lines

**Review by DevScyu — COMMENTED** [review]
Please remove empty lines throughout the code.

**Review by kristofbolyai — COMMENTED** [review]
Firstly, run you changed code through some code formatter. You have a lot of stylistic issues I did not point out 1 by 1.

Secondly, do not change info variable names. I don't mind if you add new obviously, but let's not break everyone's config.

**Comment by kristofbolyai** [comment]
> Also i would like to be a Wynntils Developper if its possible

You have to contribute enough so you stand out and you get an internship invitation.

File: src/main/java/com/wynntils/core/utils/StringUtils.java (hunk 7 above)
Comment (Incompleteusern) [reply src/main/java/com/wynntils/core/utils/StringUtils.java]: I don't think the valueOf is even needed either

File: src/main/java/com/wynntils/modules/utilities/instances/InfoFormatter.java (hunk 1 above)
Comment (Incompleteusern) [reply src/main/java/com/wynntils/modules/utilities/instances/InfoFormatter.java]: Closing as duplicate of scyu's comment

File: src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java (hunk 2 above)
Comment (Incompleteusern) [reply src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java]: Same comment

**Comment by kristofbolyai** [comment]
Also, please rename your PR. You are adding info overlay and levelling overlay variables, so make sure your PR title says that.

File: src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java [hunk 10]
```diff
@@ -41,6 +42,31 @@ public static void update() {
...
+    /**
+     * Get Excepted Grind Time for leveling up
+     * @return The Grind Time for Leveling Up in Minutes
+     */
+    public static int getLevelingGrindTime(){
+        CharacterData p = PlayerInfo.get(CharacterData.class);
+        int perminute = getXpPerMinute();
+        int required = p.getXpNeededToLevelUp() - p.getCurrentXP();
+
+
```
Comment (Incompleteusern) [inline src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java]: extra space

**Comment by Incompleteusern** [comment]
The idea of the time needed to get time for grinding and time for travelling are novel, nice!

**Comment by kristofbolyai** [comment]
Other than the changes, thank you for your contribution. 

I see you want to work on "Wynntils-Reloaded" (as your fork title says). Maybe you should try to contribute to https://github.com/Wynntils/Artemis. Artemis is 1.18.2 rewrite of Wynntils, and will likely replace this version of Wynntils when the porting is finished.

File: src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java (hunk 8 above)
Comment (Zffu) [reply src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java]: sorry

**Comment by Zffu** [comment]
When i mean Wynntils Reloaded i mean the 1.12.2 version also im gonna fix the extra spaces and unnedded code

**Comment by Zffu** [comment]
i think i fixed everything

**Comment by Incompleteusern** [comment]
> i think i fixed everything

Still haven't resolved the `valueOf` comment, but everything else seems to be resolved, thanks!

**Comment by Zffu** [comment]
oh lemme fix that rq

File: src/main/java/com/wynntils/core/utils/StringUtils.java [hunk 11]
```diff
@@ -453,6 +453,20 @@ public static String integerToShortString(int count) {
         return String.format("%s%c", value, "kMBTPE".charAt(exp - 1));
     }
 
+    public static String durationIntegerToShortString(int count){
+        if(count < 60){
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/core/utils/StringUtils.java]:
```suggestion
        if (count < 60){
```

File: src/main/java/com/wynntils/core/utils/StringUtils.java [hunk 12]
```diff
@@ -453,6 +453,20 @@ public static String integerToShortString(int count) {
...
+            return "" + count + "s";
+        }
+        else {
+            if(count > 3600){
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/core/utils/StringUtils.java]:
```suggestion
            if (count > 3600){
```

File: src/main/java/com/wynntils/core/utils/StringUtils.java [hunk 13]
```diff
@@ -453,6 +453,20 @@ public static String integerToShortString(int count) {
         return String.format("%s%c", value, "kMBTPE".charAt(exp - 1));
     }
 
+    public static String durationIntegerToShortString(int count){
+        if(count < 60){
+            return "" + count + "s";
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/core/utils/StringUtils.java]: remove "" +

File: src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java [hunk 14]
```diff
@@ -41,6 +42,22 @@ public static void update() {
...
 
+    /**
+     * Get Excepted Grind Time for leveling up
+     * @return The Grind Time for Leveling Up in Minutes
+     */
+    public static int getLevelingGrindTime(){
+        CharacterData p = PlayerInfo.get(CharacterData.class);
+        int perminute = getXpPerMinute();
+        int required = p.getXpNeededToLevelUp() - p.getCurrentXP();
+        if(perminute != 0){
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java]:
```suggestion
        if (perminute != 0){
```

**Comment by Zffu** [comment]
i removed the "" at the start of the return

File: src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java [hunk 15]
```diff
@@ -20,4 +22,23 @@ public static double getCurrentSpeed() {
...
     }
 
+    public static int getTravelingTime(){
+        Location compass = CompassManager.getCompassLocation();
+        if(compass != null){
+            Location playerPos = new Location(McIf.player());
+            double distance = compass.distance(playerPos);
+            double bps = getCurrentSpeed();
+            if(bps != 0){
+                return (int) ((int)distance / bps);
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java]: Why the double int cast?

Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java]:
```suggestion
                return (int) (distance / bps);
```

File: src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java [hunk 16]
```diff
@@ -20,4 +22,23 @@ public static double getCurrentSpeed() {
         return (MathHelper.sqrt((distX * distX) + (distZ * distZ))) * 20d;
     }
 
+    public static int getTravelingTime(){
+        Location compass = CompassManager.getCompassLocation();
+        if(compass != null){
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java]:
```suggestion
        if (compass != null) {
```

File: src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java [hunk 17]
```diff
@@ -20,4 +22,23 @@ public static double getCurrentSpeed() {
         return (MathHelper.sqrt((distX * distX) + (distZ * distZ))) * 20d;
     }
 
+    public static int getTravelingTime(){
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java]:
```suggestion
    public static int getTravelingTime() {
```

File: src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java [hunk 18]
```diff
@@ -20,4 +22,23 @@ public static double getCurrentSpeed() {
...
+            }
+            else {
+                return 0;
+            }
+
+        }
+        else {
+            return 0;
+        }
+    }
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java]: You can remove these elses and return 0 at the end of the method.

File: src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java [hunk 19]
```diff
@@ -134,7 +134,8 @@ public void onClientTick(TickEvent.ClientTickEvent e) {
                                         .replace("%xc%", Integer.toString(data.getCurrentXP() - oldxp))
                                         .replace("%po%", oldxppercent)
                                         .replace("%pn%", data.getCurrentXPAsPercentage())
-                                        .replace("%pc%", df.format(xpchange)));
+                                        .replace("%pc%", com.wynntils.core.utils.StringUtils.integerToShortString((int) xpchange))
+                                        .replace("%pc_raw%", df.format(xpchange)));
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java]: You did not revert the changes. I asked you to not change old variables, but add new.

File: src/main/java/com/wynntils/modules/utilities/overlays/hud/LevelingOverlay.java [hunk 20]
```diff
@@ -41,15 +43,19 @@ public void tick(ClientTickEvent event, long ticks) {
...
+                    .replace("%actual%", "" + StringUtils.integerToShortString(data.getCurrentXP()))
                     .replace("%percent%", data.getCurrentXPAsPercentage())
-                    .replace("%needed%", "" + (data.getXpNeededToLevelUp() - data.getCurrentXP()))
-                    .replace("%actualg%", GROUPED_FORMAT.format(data.getCurrentXP()))
-                    .replace("%maxg%", GROUPED_FORMAT.format(data.getXpNeededToLevelUp()))
-                    .replace("%neededg%", GROUPED_FORMAT.format(data.getXpNeededToLevelUp() - data.getCurrentXP()))
+                    .replace("%needed_raw%", "" + (data.getXpNeededToLevelUp() - data.getCurrentXP()))
+                    .replace("%actual_rawg%", GROUPED_FORMAT.format(data.getCurrentXP()))
+                    .replace("%max_rawg%", GROUPED_FORMAT.format(data.getXpNeededToLevelUp()))
+                    .replace("%needed_rawg%", GROUPED_FORMAT.format(data.getXpNeededToLevelUp() - data.getCurrentXP()))
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/overlays/hud/LevelingOverlay.java]: Revert these too.

File: src/main/java/com/wynntils/modules/utilities/overlays/hud/LevelingOverlay.java [hunk 21]
```diff
@@ -41,15 +43,19 @@ public void tick(ClientTickEvent event, long ticks) {
     public void render(RenderGameOverlayEvent.Pre event) {
         CharacterData data = get(CharacterData.class);
         if (((event.getType() == RenderGameOverlayEvent.ElementType.EXPERIENCE) || (event.getType() == RenderGameOverlayEvent.ElementType.JUMPBAR)) && Reference.onWorld && data.isLoaded()) {
-            String text = OverlayConfig.Leveling.INSTANCE.levelingText.replace("%actual%", "" + data.getCurrentXP())
-                    .replace("%max%", "" + data.getXpNeededToLevelUp())
+            String text = OverlayConfig.Leveling.INSTANCE.levelingText.replace("%actual%_raw", "" + data.getCurrentXP())
+                    .replace("%max_raw%", "" + data.getXpNeededToLevelUp())
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/overlays/hud/LevelingOverlay.java]: This too

**Review by kristofbolyai — CHANGES_REQUESTED** [review]

**Comment by Zffu** [comment]
ok

**Comment by Zffu** [comment]
i can rename the new ones to %max_formated% and %actual_formated% ?

**Comment by kristofbolyai** [comment]
> i can rename the new ones to %max_formated% and %actual_formated% ?

Sure. That way old variables will function the same, but there will be new, formatted ones.

**Comment by Zffu** [comment]
done

**Comment by Zffu** [comment]
fixed the elses and other things

**Comment by kristofbolyai** [comment]
@Radi0o check open conversations

**Comment by Zffu** [comment]
i resolved them

**Comment by Zffu** [comment]
i gtg rn i will add more things tomorrow i already have some cool ideas

File: src/main/java/com/wynntils/core/utils/StringUtils.java [hunk 22]
```diff
@@ -453,6 +453,20 @@ public static String integerToShortString(int count) {
         return String.format("%s%c", value, "kMBTPE".charAt(exp - 1));
     }
 
+    public static String durationIntegerToShortString(int count){
+        if (count < 60){
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/core/utils/StringUtils.java]:
```suggestion
        if (count < 60) {
```

File: src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java [hunk 23]
```diff
@@ -41,6 +42,22 @@ public static void update() {
...
 
+    /**
+     * Get Excepted Grind Time for leveling up
+     * @return The Grind Time for Leveling Up in Minutes
+     */
+    public static int getLevelingGrindTime() {
+        CharacterData p = PlayerInfo.get(CharacterData.class);
+        int perminute = getXpPerMinute();
+        int required = p.getXpNeededToLevelUp() - p.getCurrentXP();
+        if(perminute != 0){
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java]:
```suggestion
        if (perminute != 0) {
```

File: src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java [hunk 24]
```diff
@@ -20,4 +22,16 @@ public static double getCurrentSpeed() {
         return (MathHelper.sqrt((distX * distX) + (distZ * distZ))) * 20d;
     }
 
+    public static int getTravelingTime() {
+        Location compass = CompassManager.getCompassLocation();
+        if(compass != null){
+            Location playerPos = new Location(McIf.player());
+            double distance = compass.distance(playerPos);
+            double bps = getCurrentSpeed();
+            if(bps != 0){
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java]:
```suggestion
            if (bps != 0) {
```

File: src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java [hunk 25]
```diff
@@ -20,4 +22,16 @@ public static double getCurrentSpeed() {
         return (MathHelper.sqrt((distX * distX) + (distZ * distZ))) * 20d;
     }
 
+    public static int getTravelingTime() {
+        Location compass = CompassManager.getCompassLocation();
+        if(compass != null){
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/utilities/managers/SpeedometerManager.java]:
```suggestion
        if (compass != null) {
```

File: src/main/java/com/wynntils/core/utils/StringUtils.java [hunk 26]
```diff
@@ -453,6 +453,20 @@ public static String integerToShortString(int count) {
...
+            return "" + count + "s";
+        }
+        else {
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/core/utils/StringUtils.java]: Please inline your `else` clauses with the closing bracket of the previous `if` clause to be consistent with the rest of our code.
`} else {`

File: src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java [hunk 27]
```diff
@@ -3,6 +3,7 @@
 import com.wynntils.core.framework.instances.PlayerInfo;
 import com.wynntils.core.framework.instances.data.CharacterData;
 import com.wynntils.core.utils.objects.TimedSet;
+import ibxm.Player;
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java]: Unused import?

File: src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java [hunk 28]
```diff
@@ -41,6 +42,22 @@ public static void update() {
         XP_SET.releaseEntries();
     }
 
+    /**
+     * Get Excepted Grind Time for leveling up
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java]: Excepted or expected?
```suggestion
     * Get expected grind time for leveling up
```

File: src/main/java/com/wynntils/modules/utilities/overlays/hud/LevelingOverlay.java [hunk 29]
```diff
@@ -43,13 +45,17 @@ public void render(RenderGameOverlayEvent.Pre event) {
         if (((event.getType() == RenderGameOverlayEvent.ElementType.EXPERIENCE) || (event.getType() == RenderGameOverlayEvent.ElementType.JUMPBAR)) && Reference.onWorld && data.isLoaded()) {
             String text = OverlayConfig.Leveling.INSTANCE.levelingText.replace("%actual%", "" + data.getCurrentXP())
                     .replace("%max%", "" + data.getXpNeededToLevelUp())
+                    .replace("%actual_formated%", StringUtils.integerToShortString(data.getCurrentXP()))
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/utilities/overlays/hud/LevelingOverlay.java]: Please fix the typo `formated` to `formatted`. There are a couple more instances of this a few lines down.

**Review by ryanzhoudev — CHANGES_REQUESTED** [review]

**Comment by Zffu** [comment]
fixed

File: src/main/java/com/wynntils/core/utils/StringUtils.java [hunk 30]
```diff
@@ -453,6 +453,18 @@ public static String integerToShortString(int count) {
         return String.format("%s%c", value, "kMBTPE".charAt(exp - 1));
     }
 
+    public static String durationIntegerToShortString(int count){
+        if (count < 60) {
+            return "" + count + "s";
+       } else {
+            if (count > 3600){
+                return count / 3600 + "h";
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/core/utils/StringUtils.java]: maybe switch to `else if`?

**Comment by Zffu** [comment]
Done fixing the typos

**Comment by Zffu** [comment]
and the else if

**Review by Zffu — COMMENTED** [review]
fixed

**Comment by Zffu** [comment]
@DonkeyBlaster if i have more variables to add did i need to make a new pull request?

**Comment by kristofbolyai** [comment]
> @DonkeyBlaster if i have more variables to add did i need to make a new pull request?

If you don't stop updating PRs then it is never gonna get merged (change requests don't count). You should make a new PR.

**Comment by kristofbolyai** [comment]
Generally you should only open a PR if you are happy with the changes, it works, and it is semi-final version.

File: src/main/java/com/wynntils/core/utils/StringUtils.java [hunk 31]
```diff
@@ -453,6 +453,18 @@ public static String integerToShortString(int count) {
...
+        if (count < 60) {
+            return count + "s";
+       } else if (count > 86400){
+                return count / 86400 + "d";
+             } else if(count > 3600) {
+            return count / 3600 + "h";
+        } else {
+            return count / 60 + "m";
+        }
+        }
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/core/utils/StringUtils.java]: Again the whole method formatting is a mess. Format it

File: src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java [hunk 32]
```diff
@@ -8,7 +8,7 @@
 import java.util.concurrent.TimeUnit;
 
 /**
- * Idea: Maybe add a "Time Left" option to see how long leveling up takes at the current XP rate?
+ * Idea: Maybe add a "Time Left" option to see how long leveling up takes at the current XP rate? Done
  */
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/LevelingManager.java]: Let's just remove this whole comment imo

**Comment by Zffu** [comment]
ok

File: src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java (hunk 19 above)
Comment (kristofbolyai) [reply src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java]: Why did you mark this completed? You changed what %pc% does and added it back as %pc_raw%.

**Review by kristofbolyai — CHANGES_REQUESTED** [review]

**Comment by Zffu** [comment]
oh i forgot to change this one

File: src/main/java/com/wynntils/core/utils/StringUtils.java (hunk 31 above)
Comment (Zffu) [reply src/main/java/com/wynntils/core/utils/StringUtils.java]: ok

**Comment by Zffu** [comment]
i formatted the StringUtils function and reverted the %pc% variable

File: src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java [hunk 33]
```diff
@@ -123,6 +123,7 @@ public void onClientTick(TickEvent.ClientTickEvent e) {
                                         .replace("%xc%", Integer.toString(data.getCurrentXP() - oldxp))
                                         .replace("%po%", oldxppercent)
                                         .replace("%pn%", data.getCurrentXPAsPercentage())
+                                        .replace("%pc_formatted%", com.wynntils.core.utils.StringUtils.integerToShortString((int) xpchange))
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java]: Please import this properly

File: src/main/java/com/wynntils/modules/utilities/overlays/hud/LevelingOverlay.java [hunk 34]
```diff
@@ -43,13 +45,17 @@ public void render(RenderGameOverlayEvent.Pre event) {
...
                     .replace("%percent%", data.getCurrentXPAsPercentage())
                     .replace("%needed%", "" + (data.getXpNeededToLevelUp() - data.getCurrentXP()))
                     .replace("%actualg%", GROUPED_FORMAT.format(data.getCurrentXP()))
                     .replace("%maxg%", GROUPED_FORMAT.format(data.getXpNeededToLevelUp()))
                     .replace("%neededg%", GROUPED_FORMAT.format(data.getXpNeededToLevelUp() - data.getCurrentXP()))
+                    .replace("%needed_formatted%", StringUtils.integerToShortString(data.getXpNeededToLevelUp() - data.getCurrentXP()))
+                    .replace("%max_formatted%", StringUtils.integerToShortString(data.getXpNeededToLevelUp()))
                     .replace("%curlvl%", "" + data.getLevel())
-                    .replace("%nextlvl%", data.getLevel() == 104 ? "" : "" + (data.getLevel() + 1));
+                    .replace("%nextlvl%", data.getLevel() == 104 ? "" : "" + (data.getLevel() + 1))
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/utilities/overlays/hud/LevelingOverlay.java]:
```suggestion
                    .replace("%nextlvl%", data.getLevel() == 106 ? "" : "" + (data.getLevel() + 1))
```

**Review by ryanzhoudev — COMMENTED** [review]
Tried it in-game, seems fine

File: src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java (hunk 33 above)
Comment (Zffu) [reply src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java]: i cant there is another StringUtils class imported

Comment (Zffu) [reply src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java]: nevermind i found a way

File: src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java [hunk 35]
```diff
@@ -32,12 +32,15 @@
 import net.minecraftforge.fml.common.gameevent.TickEvent;
 import org.apache.commons.lang3.StringUtils;
 
+
```
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java]: Remove random blank line

**Pull request closed without merging**